### PR TITLE
UCT/ROCM/IPC: detect link type and declare correct BW for rocm_ipc

### PR DESCRIPTION
**Pull request by souravzzz**
## What
Detect link type (PCIe/xGMI) and declare appropriate bandwidth
Note: assumes all GPUs are connected through same type of link

## Why ?
Fix https://github.com/openucx/ucx/issues/4855

### CODE REVIEW

File: src/uct/rocm/ipc/rocm_ipc_iface.c [hunk 1]
```diff
@@ -27,6 +27,26 @@ static ucs_config_field_t uct_rocm_ipc_iface_config_table[] = {
...
+        case HSA_AMD_LINK_INFO_TYPE_PCIE:
+            bw = 200.0 * UCS_GBYTE;
+            break;
+        case HSA_AMD_LINK_INFO_TYPE_XGMI:
+            bw = 400.0 * UCS_GBYTE;
+            break;
+        default:
+            bw = 100.0 * UCS_GBYTE;
+            break;
+    }
```
Comment (yosefe) [inline src/uct/rocm/ipc/rocm_ipc_iface.c:46]: code style: switch/case on same indent, space after "switch"

File: src/uct/rocm/ipc/rocm_ipc_iface.c [hunk 2]
```diff
@@ -87,11 +107,11 @@ static ucs_status_t uct_rocm_ipc_iface_query(uct_iface_h tl_iface,
                                           UCT_IFACE_FLAG_PENDING   |
                                           UCT_IFACE_FLAG_CONNECT_TO_IFACE;
 
-    /* TODO: get accurate info */
-    iface_attr->latency                 = ucs_linear_func_make(80e-9, 0);
-    iface_attr->bandwidth.dedicated     = 10.0 * UCS_GBYTE; /* 10 GB */
-    iface_attr->bandwidth.shared        = 0;
-    iface_attr->overhead                = 0.4e-6; /* 0.4 us */
+    iface_attr->latency                 = ucs_linear_func_make(1e-9, 0);
```
Comment (yosefe) [inline src/uct/rocm/ipc/rocm_ipc_iface.c:110]: 1 nanosec seems too small?

File: src/uct/rocm/base/rocm_base.c [hunk 3]
```diff
@@ -223,6 +222,61 @@ ucs_status_t uct_rocm_base_mem_query(uct_md_h md, const void *addr,
...
+ucs_status_t uct_rocm_base_get_link_type(hsa_amd_link_info_type_t *link_type)
+{
+    hsa_amd_memory_pool_link_info_t link_info;
+    hsa_agent_t agent1, agent2;
+    hsa_amd_memory_pool_t pool;
+    hsa_status_t status;
+
+    *link_type = HSA_AMD_LINK_INFO_TYPE_PCIE;
+
+    if (uct_rocm_base_agents.num_gpu > 1) {
```
Comment (yosefe) [inline src/uct/rocm/base/rocm_base.c:256]: if (uct_rocm_base_agents.num_gpu <= 1) { 
  return UCS_OK;
}

File: src/uct/rocm/ipc/rocm_ipc_iface.c (hunk 1 above)
Comment (souravzzz) [reply src/uct/rocm/ipc/rocm_ipc_iface.c]: Fixed

File: src/uct/rocm/base/rocm_base.c (hunk 3 above)
Comment (souravzzz) [reply src/uct/rocm/base/rocm_base.c]: Fixed

File: src/uct/rocm/ipc/rocm_ipc_iface.c (hunk 2 above)
Comment (souravzzz) [reply src/uct/rocm/ipc/rocm_ipc_iface.c:110]: Was not sure how this value impacts protocol selection so decided to use the same value as https://github.com/openucx/ucx/blob/master/src/uct/cuda/cuda_ipc/cuda_ipc_iface.c#L154

**Review by yosefe — APPROVED** [review]
can you pls squash?

File: src/uct/rocm/base/rocm_base.c [hunk 4]
```diff
@@ -223,6 +222,62 @@ ucs_status_t uct_rocm_base_mem_query(uct_md_h md, const void *addr,
...
+    if (uct_rocm_base_agents.num_gpu < 2) {
+        return UCS_OK;
+    }
+
+    agent1 = uct_rocm_base_agents.gpu_agents[0];
+    agent2 = uct_rocm_base_agents.gpu_agents[1];
+
+    status = hsa_amd_agent_iterate_memory_pools(agent2,
+                            uct_rocm_hsa_pool_callback, (void*)&pool);
+    if (status != HSA_STATUS_SUCCESS && status != HSA_STATUS_INFO_BREAK) {
```
Comment (brminich) [inline src/uct/rocm/base/rocm_base.c:265]: codestyle: please use `()` for each statement:
`if ((status != HSA_STATUS_SUCCESS) && (status != HSA_STATUS_INFO_BREAK))`

Comment (souravzzz) [reply src/uct/rocm/base/rocm_base.c]: Fixed

**Review by yosefe — APPROVED** [review]

**Pull request merged**